simulation_ws/src/test_nodes/nodes/coverage_test.py

```diff
@@ -0,0 +1,188 @@
+        Overloaded unittest method to set up test variables
+        """
+        self.test_name = self.TEST_NAME + "_" + str(time.time()).split(".")[0]
+        self.is_cancelled = False
+        while self.get_time() == 0:
+            pass
+        self.sim_start_time_secs = self.get_time()
+        self.coverage_start_time = self.get_time()  # For later usage, now it does the same as sim time
+        self.percentage_covered = 0
+        self.tl = tf.TransformListener()
+        self.get_params()
+        self.get_map()
+        self.painting_map = self.map_array
+
+    def runTest(self):
+        """
+        Start the coverage test
+        """
+        try:
+            rospy.loginfo(self.test_name)
+            self.set_tag(name="Test_Start",
+                         value=str(self.sim_start_time_secs))
+            rospy.Subscriber("/clock", Clock, callback=self.timeout_cb)
+            rospy.Subscriber(self.odom_topic, Odometry, callback=self.odom_cb)
+
+            rospy.spin()
+        except Exception as ex:
+            rospy.logerr("Error %s", ex)
+            self.set_tag(name=self.test_name, value="Failed")
+            self.cancel_job()
+
+
+    def odom_cb(self, msg):
+        pose_in_odom = self.odometry_to_pose_stamped(msg)
+
+        # Getting an extrapolation error, even when asking if it canTransform
+        if self.tl.canTransform(self.map_topic, self.odom_topic, rospy.Time(0)):
+            pose_in_map = self.tl.transformPose(self.map_topic, pose_in_odom)
+
+            map_pos_in_array_x = floor((pose_in_map.pose.position.x / self.resolution) + self.map_origin_x)
+            map_pos_in_array_y = floor((pose_in_map.pose.position.y / self.resolution) + self.map_origin_y)
+            norm_map_pos_in_array_x = int(round(map_pos_in_array_x + self.map_height/2))
+            norm_map_pos_in_array_y = int(round(map_pos_in_array_y + self.map_width/2))
+
+            if self.map_array[norm_map_pos_in_array_x][norm_map_pos_in_array_y] != self.COVERED_PGM_VALUE:
+                self.painting_map[norm_map_pos_in_array_x][norm_map_pos_in_array_y] = self.COVERED_PGM_VALUE
+
+            rospy.loginfo ("Original free spaces -> %s", self.map_free_spaces)
+            rospy.loginfo ("Current free spaces -> %s", np.count_nonzero(self.painting_map == self.FREE_PGM_VALUE))
+            current_percentage_covered = float((np.count_nonzero(self.painting_map == self.FREE_PGM_VALUE) / self.map_free_spaces ) * 100)
+            rospy.loginfo(current_percentage_covered)
+
+    def odometry_to_pose_stamped(self, msg):
+        pose_in_odom = PoseStamped()
+        pose_in_odom.header = msg.header
+        pose_in_odom.pose = msg.pose.pose
+        return pose_in_odom
+
+    def get_map(self):
+        """
+        Save the raw map and convert it to a numpy array. Saves useful data.
+        """
+        self.raw_map = rospy.wait_for_message(self.map_topic, OccupancyGrid, timeout=5)
+        self.map_array = np.array(self.raw_map.data)
```

Comment:
https://numpy.org/doc/stable/reference/generated/numpy.reshape.html
You know the width and height, and a 2D map would be easier to manipulate. Anyways, it's important to check that the order in which the reshape leaves rows is correct.